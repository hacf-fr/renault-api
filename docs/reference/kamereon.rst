Kamereon low-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Core
----

.. automodule:: renault_api.kamereon
   :members:

Data models
-----------

.. automodule:: renault_api.kamereon.models
   :members:
