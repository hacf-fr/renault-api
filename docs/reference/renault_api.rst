Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
