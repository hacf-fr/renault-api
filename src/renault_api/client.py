"""Client for Renault API."""


class RenaultClient:
    """Proxy to the Renault API."""

    def __init__(self) -> None:
        """Initialise Renault Client."""
