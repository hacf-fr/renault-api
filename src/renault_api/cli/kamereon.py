"""Singletons for the CLI."""
import click
from tabulate import tabulate

from .core import CLICredentialStore
from .core import CONF_ACCOUNT_ID
from .core import CONF_VIN
from .core import ensure_locale
from renault_api.const import CONF_LOCALE
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys
from renault_api.kamereon import CREDENTIAL_GIGYA_LOGIN_TOKEN
from renault_api.kamereon import Kamereon


class CLIKamereon:
    """Singleton of the CredentialStore for the CLI."""

    __instance: Kamereon = None

    @classmethod
    async def get_client(cls, websession):
        """Get singleton Kamereon client."""
        if not CLIKamereon.__instance:
            await ensure_locale(websession)

            credential_store = CLICredentialStore()
            locale = credential_store.get_value(CONF_LOCALE)
            country = locale[-2:]
            api_keys = await get_api_keys(locale, aiohttp_session=websession)
            return Kamereon(
                websession, country, api_keys, credential_store=credential_store
            )
        return CLIKamereon.__instance


async def ensure_logged_in(websession) -> None:
    """Prompt the user for credentials."""
    await ensure_locale(websession)

    credential_store = CLICredentialStore()
    if CREDENTIAL_GIGYA_LOGIN_TOKEN in credential_store:
        return

    while True:
        user = click.prompt("user")
        password = click.prompt("password", hide_input=True)
        try:
            await do_login(websession, user, password)
        except RenaultException as exc:
            click.echo(f"Login failed: {exc}.", err=True)
        else:
            return


async def do_login(websession, user, password):
    """Attempt login."""
    await ensure_locale(websession)
    kamereon = await CLIKamereon.get_client(websession)
    await kamereon.login(user, password)


async def get_account(websession) -> str:
    """Prompt the user for account."""
    credential_store = CLICredentialStore()
    if CONF_ACCOUNT_ID in credential_store:
        return credential_store.get_value(CONF_ACCOUNT_ID)

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    if not response.accounts:
        raise RenaultException("No account found.")
    if len(response.accounts) == 1:
        return response.accounts[0].accountId
    elif len(response.accounts) > 1:
        menu = "Multiple accounts found:\n"
        for i, account in enumerate(response.accounts):
            menu = menu + \
                f"\t[{i+1}] {account.accountId} ({account.accountType})\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                account = response.accounts[i - 1].accountId
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)
            else:
                click.echo("To avoid seeing this message again, use:"
                           f"> renault-api set --account {account}")
                return account


async def get_vin(websession, account):
    """Prompt the user for vin."""
    credential_store = CLICredentialStore()
    if CONF_VIN in credential_store:
        return

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_vehicles(account)
    if not response.vehicleLinks:
        raise RenaultException("No vehicle found.")
    if len(response.vehicleLinks) == 1:
        return response.vehicleLinks[0].vin
    elif len(response.vehicleLinks) > 1:
        menu = "Multiple vehicles found:\n"
        for i, vehicle in enumerate(response.vehicleLinks):
            menu = menu + f"\t[{i+1}] {vehicle.vin}\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                return response.vehicleLinks[i - 1].vin
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)


async def display_accounts(websession) -> None:
    """Login to Renault."""
    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    accounts = {
        account.accountType: account.accountId for account in response.accounts}
    click.echo(tabulate(accounts.items(), headers=[
               "Type", "ID"]))
