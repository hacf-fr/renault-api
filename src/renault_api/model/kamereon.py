"""Kamereon models."""
import json
from dataclasses import dataclass
from typing import List
from typing import Optional

import marshmallow_dataclass

from . import BaseModel
from . import BaseSchema
from renault_api.exceptions import KamereonResponseException


@dataclass
class KamereonResponseError(BaseModel):
    """Kamereon response error."""

    errorCode: Optional[str]  # noqa: N815
    errorMessage: Optional[str]  # noqa: N815

    def raise_for_error_code(self) -> None:
        """Raise exception from response error."""
        raise KamereonResponseException(self.errorCode, self.get_error_details())

    def get_error_details(self) -> Optional[str]:
        """Extract the error details sometimes hidden inside nested JSON."""
        try:
            error_details = json.loads(self.errorMessage or "{}")
        except json.JSONDecodeError:
            return self.errorMessage

        error_descriptions = []
        for inner_error in error_details.get("errors", []):
            error_description = " ".join(
                filter(
                    None,
                    [
                        inner_error.get("title"),
                        inner_error.get("source", {}).get("pointer"),
                        inner_error.get("detail"),
                    ],
                )
            )
            error_descriptions.append(error_description)

        return ", ".join(error_descriptions) or self.errorMessage


@dataclass
class KamereonResponse(BaseModel):
    """Kamereon response."""

    errors: Optional[List[KamereonResponseError]]

    def raise_for_error_code(self) -> None:
        """Raise exception if errors found in the response."""
        for error in self.errors or []:
            # Until we have a sample for multiple errors, just raise on first one
            error.raise_for_error_code()


@dataclass
class KamereonPersonAccount(BaseModel):
    """Kamereon account data."""

    accountId: Optional[str]  # noqa: N815
    accountType: Optional[str]  # noqa: N815
    accountStatus: Optional[str]  # noqa: N815


@dataclass
class KamereonPersonResponse(KamereonResponse):
    """Kamereon response to GET on /persons/{gigya_person_id}."""

    accounts: List[KamereonPersonAccount]


@dataclass
class KamereonVehiclesLink(BaseModel):
    """Kamereon account data."""

    vin: Optional[str]


@dataclass
class KamereonVehiclesResponse(KamereonResponse):
    """Kamereon response to GET on /accounts/{account_id}/vehicles."""

    accountId: Optional[str]  # noqa: N815
    country: Optional[str]
    vehicleLinks: List[KamereonVehiclesLink]  # noqa: N815


@dataclass
class KamereonVehicleData(BaseModel):
    """Kamereon account data."""

    type: Optional[str]
    id: Optional[str]


@dataclass
class KamereonVehicleDataResponse(KamereonResponse):
    """Kamereon response to GET/POST on .../cars/{vin}/{type}."""

    data: Optional[KamereonVehicleData]


KamereonPersonResponseSchema = marshmallow_dataclass.class_schema(
    KamereonPersonResponse, base_schema=BaseSchema
)()
KamereonVehiclesResponseSchema = marshmallow_dataclass.class_schema(
    KamereonVehiclesResponse, base_schema=BaseSchema
)()
KamereonVehicleDataResponseSchema = marshmallow_dataclass.class_schema(
    KamereonVehicleDataResponse, base_schema=BaseSchema
)()
