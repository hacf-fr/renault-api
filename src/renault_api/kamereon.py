"""Kamereon client for interaction with Renault servers."""
import logging
from typing import Any
from typing import cast
from typing import Dict
from typing import Optional

from aiohttp import ClientSession
from marshmallow.schema import Schema

from .const import CONF_KAMEREON_APIKEY
from .const import CONF_KAMEREON_URL
from .exceptions import KamereonException
from .gigya import Gigya
from renault_api.model import kamereon as model
from renault_api.model.credential import Credential
from renault_api.model.credential import JWTCredential

CREDENTIAL_GIGYA_JWT = "gigya_jwt"
CREDENTIAL_GIGYA_LOGIN_TOKEN = "gigya_login_token"
CREDENTIAL_GIGYA_PERSON_ID = "gigya_person_id"

_LOGGER = logging.getLogger(__name__)


class Kamereon:
    """Kamereon client for interaction with Renault servers."""

    def __init__(
        self,
        websession: ClientSession,
        country: str,
        locale_details: Dict[str, str],
        gigya: Optional[Gigya] = None,
    ) -> None:
        """Initialise Kamereon."""
        self._websession = websession

        self._country = country
        self._api_key = locale_details[CONF_KAMEREON_APIKEY]
        self._root_url = locale_details[CONF_KAMEREON_URL]

        self._gigya = gigya or Gigya(
            websession=websession, locale_details=locale_details
        )
        self._credentials: Dict[str, Credential] = {}

    async def _get_credential(self, key: str) -> str:
        credential = self._credentials.get(key)
        if credential and not credential.has_expired():
            return credential.value

        if key == CREDENTIAL_GIGYA_JWT:
            login_token = await self._get_credential(CREDENTIAL_GIGYA_LOGIN_TOKEN)
            jwt_response = await self._gigya.get_jwt(login_token)
            credential = JWTCredential(jwt_response.get_jwt_token())
            self._credentials[key] = credential
            return credential.value
        if key == CREDENTIAL_GIGYA_PERSON_ID:
            login_token = await self._get_credential(CREDENTIAL_GIGYA_LOGIN_TOKEN)
            account_response = await self._gigya.get_account_info(login_token)
            credential = Credential(account_response.get_person_id())
            self._credentials[key] = credential
            return credential.value

        raise KamereonException(f"Credential `{key}` not found in credential cache.")

    def _get_path_to_account(self, account_id: str) -> str:
        """Get the path to the account."""
        return f"/accounts/{account_id}"

    def _get_path_to_car(self, account_id: str, version: int, vin: str) -> str:
        """Get the path to the car."""
        account_path = self._get_path_to_account(account_id)
        return f"{account_path}/kamereon/kca/car-adapter/v{version}/cars/{vin}"

    async def _request(
        self,
        schema: Schema,
        method: str,
        path: str,
        params: Optional[Dict[str, str]] = None,
        data: Optional[Dict[str, Any]] = None,
    ) -> model.KamereonResponse:
        url = f"{self._root_url}/commerce/v1{path}"
        headers = {
            "apikey": self._api_key,
            "x-gigya-id_token": await self._get_credential(CREDENTIAL_GIGYA_JWT),
        }
        params = params or {}
        params["country"] = self._country

        async with self._websession.request(
            method,
            url,
            headers=headers,
            params=params,
            data=data,
        ) as http_response:
            response_text = await http_response.text()
            _LOGGER.debug(
                "Received Kamereon response %s on %s: %s",
                http_response.status,
                http_response.url,
                response_text,
            )
            kamereon_response: model.KamereonResponse = schema.loads(response_text)
            # Check for Kamereon error
            kamereon_response.raise_for_error_code()
            # Check for HTTP error
            http_response.raise_for_status()

            return kamereon_response

    async def login(self, login_id: str, password: str) -> None:
        """Forward login to Gigya, and cache the login token."""
        self._credentials.clear()
        response = await self._gigya.login(login_id, password)
        self._credentials[CREDENTIAL_GIGYA_LOGIN_TOKEN] = Credential(
            response.get_session_cookie()
        )

    async def get_person(self) -> model.KamereonPersonResponse:
        """GET to /persons/{person_id}."""
        person_id = await self._get_credential(CREDENTIAL_GIGYA_PERSON_ID)
        return cast(
            model.KamereonPersonResponse,
            await self._request(
                schema=model.KamereonPersonResponseSchema,
                method="GET",
                path=f"/persons/{person_id}",
            ),
        )

    async def get_vehicles(self, account_id: str) -> model.KamereonVehiclesResponse:
        """GET to /accounts/{account_id}/vehicles."""
        return cast(
            model.KamereonVehiclesResponse,
            await self._request(
                schema=model.KamereonVehiclesResponseSchema,
                method="GET",
                path=f"/accounts/{account_id}/vehicles",
            ),
        )

    async def get_vehicle_data(
        self,
        account_id: str,
        endpoint_version: int,
        vin: str,
        endpoint: str,
        params: Optional[Dict[str, str]] = None,
    ) -> model.KamereonVehicleDataResponse:
        """GET to /v{endpoint_version}/cars/{vin}/{endpoint}."""
        path_to_car = self._get_path_to_car(account_id, endpoint_version, vin)
        return cast(
            model.KamereonVehicleDataResponse,
            await self._request(
                schema=model.KamereonVehicleDataResponseSchema,
                method="GET",
                path=f"{path_to_car}/{endpoint}",
                params=params,
            ),
        )

    async def get_vehicle_battery_status(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/battery-status."""
        return await self.get_vehicle_data(account_id, 2, vin, "battery-status")

    async def get_vehicle_location(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/location."""
        return await self.get_vehicle_data(account_id, 1, vin, "location")

    async def get_vehicle_hvac_status(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/hvac-status."""
        return await self.get_vehicle_data(account_id, 1, vin, "hvac-status")

    async def get_vehicle_charge_mode(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/charge-mode."""
        return await self.get_vehicle_data(account_id, 1, vin, "charge-mode")

    async def get_vehicle_cockpit(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/cockpit."""
        return await self.get_vehicle_data(account_id, 2, vin, "cockpit")

    async def get_vehicle_lock_status(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/lock-status."""
        return await self.get_vehicle_data(account_id, 1, vin, "lock-status")

    async def get_vehicle_charging_settings(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/charging-settings."""
        return await self.get_vehicle_data(account_id, 1, vin, "charging-settings")

    async def get_vehicle_notification_settings(
        self, account_id: str, vin: str
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/notification-settings."""
        return await self.get_vehicle_data(account_id, 1, vin, "notification-settings")

    async def get_vehicle_charges(
        self, account_id: str, vin: str, params: Dict[str, str]
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/charges."""
        return await self.get_vehicle_data(account_id, 1, vin, "charges", params)

    async def get_vehicle_charge_history(
        self, account_id: str, vin: str, params: Dict[str, str]
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/charge-history."""
        return await self.get_vehicle_data(account_id, 1, vin, "charge-history", params)

    async def get_vehicle_hvac_sessions(
        self, account_id: str, vin: str, params: Dict[str, str]
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/hvac-sessions."""
        return await self.get_vehicle_data(account_id, 1, vin, "hvac-sessions", params)

    async def get_vehicle_hvac_history(
        self, account_id: str, vin: str, params: Dict[str, str]
    ) -> model.KamereonVehicleDataResponse:
        """GET to /cars/{vin}/hvac-history."""
        return await self.get_vehicle_data(account_id, 1, vin, "hvac-history", params)

    async def set_vehicle_data(
        self,
        account_id: str,
        endpoint_version: int,
        vin: str,
        endpoint: str,
        data: Dict[str, Any],
    ) -> model.KamereonVehicleDataResponse:
        """POST to /v{endpoint_version}/cars/{vin}/actions/{endpoint}."""
        path_to_car = self._get_path_to_car(account_id, endpoint_version, vin)
        return cast(
            model.KamereonVehicleDataResponse,
            await self._request(
                schema=model.KamereonVehicleDataResponseSchema,
                method="POST",
                path=f"{path_to_car}/actions/{endpoint}",
                data=data,
            ),
        )

    async def set_vehicle_hvac_start(
        self, account_id: str, vin: str, attributes: Dict[str, Any]
    ) -> model.KamereonVehicleDataResponse:
        """POST to /cars/{vin}/actions/hvac-start."""
        data = {"type": "HvacStart", "attributes": attributes}
        return await self.set_vehicle_data(account_id, 1, vin, "hvac-start", data)

    async def set_vehicle_charge_schedule(
        self, account_id: str, vin: str, attributes: Dict[str, Any]
    ) -> model.KamereonVehicleDataResponse:
        """POST to /cars/{vin}/actions/charge-schedule."""
        data = {"type": "ChargeSchedule", "attributes": attributes}
        return await self.set_vehicle_data(account_id, 2, vin, "charge-schedule", data)

    async def set_vehicle_charge_mode(
        self, account_id: str, vin: str, attributes: Dict[str, Any]
    ) -> model.KamereonVehicleDataResponse:
        """POST to /cars/{vin}/actions/charge-mode."""
        data = {"type": "ChargeMode", "attributes": attributes}
        return await self.set_vehicle_data(account_id, 1, vin, "charge-mode", data)

    async def set_vehicle_charging_start(
        self, account_id: str, vin: str, attributes: Dict[str, Any]
    ) -> model.KamereonVehicleDataResponse:
        """POST to /cars/{vin}/actions/charging-start."""
        data = {"type": "ChargingStart", "attributes": attributes}
        return await self.set_vehicle_data(account_id, 1, vin, "charging-start", data)
