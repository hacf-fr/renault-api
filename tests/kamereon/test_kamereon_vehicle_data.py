"""Tests for Kamereon models."""
from typing import cast

import pytest
from tests import fixtures

from renault_api.kamereon import enums
from renault_api.kamereon import models
from renault_api.kamereon import schemas


@pytest.mark.parametrize(
    "filename",
    fixtures.get_json_files(f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data"),
)
def test_vehicle_data_response(filename: str) -> None:
    """Test vehicle data response."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        filename, schemas.KamereonVehicleDataResponseSchema
    )
    response.raise_for_error_code()
    # Ensure the VIN is hidden
    assert response.data.id.startswith("VF1AAAA")


def test_battery_status_1() -> None:
    """Test vehicle data for battery-status.1.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/battery-status.1.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {
        "timestamp": "2020-11-17T09:06:48+01:00",
        "batteryLevel": 50,
        "batteryAutonomy": 128,
        "batteryCapacity": 0,
        "batteryAvailableEnergy": 0,
        "plugStatus": 0,
        "chargingStatus": -1.0,
    }

    vehicle_data = cast(
        models.KamereonVehicleBatteryStatusData,
        response.get_attributes(schemas.KamereonVehicleBatteryStatusDataSchema),
    )

    assert vehicle_data.timestamp == "2020-11-17T09:06:48+01:00"
    assert vehicle_data.batteryLevel == 50
    assert vehicle_data.batteryTemperature is None
    assert vehicle_data.batteryAutonomy == 128
    assert vehicle_data.batteryCapacity == 0
    assert vehicle_data.batteryAvailableEnergy == 0
    assert vehicle_data.plugStatus == 0
    assert vehicle_data.chargingStatus == -1.0
    assert vehicle_data.chargingRemainingTime is None
    assert vehicle_data.chargingInstantaneousPower is None
    assert vehicle_data.get_plug_status() == enums.PlugState.UNPLUGGED
    assert vehicle_data.get_charging_status() == enums.ChargeState.CHARGE_ERROR


def test_battery_status_2() -> None:
    """Test vehicle data for battery-status.2.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/battery-status.2.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {
        "timestamp": "2020-01-12T21:40:16Z",
        "batteryLevel": 60,
        "batteryTemperature": 20,
        "batteryAutonomy": 141,
        "batteryCapacity": 0,
        "batteryAvailableEnergy": 31,
        "plugStatus": 1,
        "chargingStatus": 1.0,
        "chargingRemainingTime": 145,
        "chargingInstantaneousPower": 27.0,
    }

    vehicle_data = cast(
        models.KamereonVehicleBatteryStatusData,
        response.get_attributes(schemas.KamereonVehicleBatteryStatusDataSchema),
    )

    assert vehicle_data.timestamp == "2020-01-12T21:40:16Z"
    assert vehicle_data.batteryLevel == 60
    assert vehicle_data.batteryTemperature == 20
    assert vehicle_data.batteryAutonomy == 141
    assert vehicle_data.batteryCapacity == 0
    assert vehicle_data.batteryAvailableEnergy == 31
    assert vehicle_data.plugStatus == 1
    assert vehicle_data.chargingStatus == 1.0
    assert vehicle_data.chargingRemainingTime == 145
    assert vehicle_data.chargingInstantaneousPower == 27.0
    assert vehicle_data.get_plug_status() == enums.PlugState.PLUGGED
    assert vehicle_data.get_charging_status() == enums.ChargeState.CHARGE_IN_PROGRESS


def test_cockpit_zoe() -> None:
    """Test vehicle data for cockpit.zoe.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/cockpit.zoe.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {"totalMileage": 49114.27}

    vehicle_data = cast(
        models.KamereonVehicleCockpitData,
        response.get_attributes(schemas.KamereonVehicleCockpitDataSchema),
    )

    assert vehicle_data.totalMileage == 49114.27
    assert vehicle_data.fuelAutonomy is None
    assert vehicle_data.fuelQuantity is None


def test_cockpit_captur_ii() -> None:
    """Test vehicle data for cockpit.captur_ii.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/cockpit.captur_ii.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {
        "fuelAutonomy": 35.0,
        "fuelQuantity": 3.0,
        "totalMileage": 5566.78,
    }

    vehicle_data = cast(
        models.KamereonVehicleCockpitData,
        response.get_attributes(schemas.KamereonVehicleCockpitDataSchema),
    )

    assert vehicle_data.totalMileage == 5566.78
    assert vehicle_data.fuelAutonomy == 35.0
    assert vehicle_data.fuelQuantity == 3.0


def test_charging_settings() -> None:
    """Test vehicle data for charging-settings.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/charging-settings.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {
        "mode": "scheduled",
        "schedules": [
            {
                "id": 1,
                "activated": True,
                "monday": {"startTime": "T12:00Z", "duration": 15},
                "tuesday": {"startTime": "T04:30Z", "duration": 420},
                "wednesday": {"startTime": "T22:30Z", "duration": 420},
                "thursday": {"startTime": "T22:00Z", "duration": 420},
                "friday": {"startTime": "T12:15Z", "duration": 15},
                "saturday": {"startTime": "T12:30Z", "duration": 30},
                "sunday": {"startTime": "T12:45Z", "duration": 45},
            }
        ],
    }

    vehicle_data = cast(
        models.KamereonVehicleChargingSettingsData,
        response.get_attributes(schemas.KamereonVehicleChargingSettingsDataSchema),
    )

    assert vehicle_data.mode == "scheduled"
    assert len(vehicle_data.schedules) == 1

    schedule_data = vehicle_data.schedules[0]
    assert schedule_data.id == 1
    assert schedule_data.activated is True
    assert schedule_data.monday.startTime == "T12:00Z"
    assert schedule_data.monday.duration == 15
    assert schedule_data.tuesday.startTime == "T04:30Z"
    assert schedule_data.tuesday.duration == 420
    assert schedule_data.wednesday.startTime == "T22:30Z"
    assert schedule_data.wednesday.duration == 420
    assert schedule_data.thursday.startTime == "T22:00Z"
    assert schedule_data.thursday.duration == 420
    assert schedule_data.friday.startTime == "T12:15Z"
    assert schedule_data.friday.duration == 15
    assert schedule_data.saturday.startTime == "T12:30Z"
    assert schedule_data.saturday.duration == 30
    assert schedule_data.sunday.startTime == "T12:45Z"
    assert schedule_data.sunday.duration == 45


def test_location() -> None:
    """Test vehicle data for location.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/location.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {
        "gpsLatitude": 48.1234567,
        "gpsLongitude": 11.1234567,
        "lastUpdateTime": "2020-02-18T16:58:38Z",
    }

    vehicle_data = cast(
        models.KamereonVehicleLocationData,
        response.get_attributes(schemas.KamereonVehicleLocationDataSchema),
    )

    assert vehicle_data.gpsLatitude == 48.1234567
    assert vehicle_data.gpsLongitude == 11.1234567
    assert vehicle_data.lastUpdateTime == "2020-02-18T16:58:38Z"


def test_charge_mode() -> None:
    """Test vehicle data for charge-mode.json."""
    response: models.KamereonVehicleDataResponse = fixtures.get_file_content_as_schema(
        f"{fixtures.KAMEREON_FIXTURE_PATH}/vehicle_data/charge-mode.json",
        schemas.KamereonVehicleDataResponseSchema,
    )
    response.raise_for_error_code()
    assert response.data.raw_data["attributes"] == {"chargeMode": "always"}

    vehicle_data = cast(
        models.KamereonVehicleChargeModeData,
        response.get_attributes(schemas.KamereonVehicleChargeModeDataSchema),
    )

    assert vehicle_data.chargeMode == "always"
    assert vehicle_data.get_charge_mode() == enums.ChargeMode.ALWAYS
