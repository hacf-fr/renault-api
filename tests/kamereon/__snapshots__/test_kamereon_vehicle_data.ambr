# serializer version: 1
# name: test_charge_schedule_single
  dict({
    'calendar': dict({
      'friday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'monday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'saturday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'sunday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'thursday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'tuesday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
      'wednesday': list([
        dict({
          'activationState': False,
          'duration': 480,
          'startTime': '2300',
        }),
      ]),
    }),
  })
# ---
# name: test_ev_settings[empty]
  dict({
    'chargeDuration': 9000,
    'chargeModeRq': 'SCHEDULED',
    'chargeTimeStart': '00:00',
    'delegatedActivated': False,
    'lastSettingsUpdateTimestamp': '2025-04-15T12:06:32.266Z',
    'preconditioningHeatedLeftSeat': False,
    'preconditioningHeatedRightSeat': False,
    'preconditioningHeatedStrgWheel': False,
    'preconditioningTemperature': 20.0,
    'programs': list([
    ]),
  })
# ---
# name: test_ev_settings[multi]
  dict({
    'chargeDuration': 600,
    'chargeModeRq': 'SCHEDULED',
    'chargeTimeStart': '09:00',
    'lastSettingsUpdateTimestamp': '2024-10-17T11:27:42.928046Z',
    'preconditioningHeatedLeftSeat': False,
    'preconditioningHeatedRightSeat': False,
    'preconditioningHeatedStrgWheel': False,
    'preconditioningTemperature': 21.0,
    'programs': list([
      dict({
        'programActivationFriday': True,
        'programActivationMonday': True,
        'programActivationSaturday': True,
        'programActivationStatus': False,
        'programActivationSunday': True,
        'programActivationThursday': True,
        'programActivationTuesday': True,
        'programActivationWednesday': True,
        'programDepartureTime': '11:00',
        'programType': 'CHARGE',
      }),
      dict({
        'programActivationFriday': True,
        'programActivationMonday': True,
        'programActivationSaturday': True,
        'programActivationStatus': False,
        'programActivationSunday': True,
        'programActivationThursday': True,
        'programActivationTuesday': True,
        'programActivationWednesday': True,
        'programDepartureTime': '13:00',
        'programType': 'CHARGE',
      }),
      dict({
        'programActivationFriday': True,
        'programActivationMonday': True,
        'programActivationSaturday': True,
        'programActivationStatus': False,
        'programActivationSunday': True,
        'programActivationThursday': True,
        'programActivationTuesday': True,
        'programActivationWednesday': True,
        'programDepartureTime': '15:00',
        'programType': 'CHARGE',
      }),
    ]),
  })
# ---
# name: test_ev_settings[single]
  dict({
    'chargeDuration': 9000,
    'chargeModeRq': 'SCHEDULED',
    'chargeTimeStart': '00:00',
    'delegatedActivated': False,
    'lastSettingsUpdateTimestamp': '2025-04-24T12:17:12.247Z',
    'preconditioningHeatedLeftSeat': False,
    'preconditioningHeatedRightSeat': False,
    'preconditioningHeatedStrgWheel': False,
    'preconditioningTemperature': 20.0,
    'programs': list([
      dict({
        'programActivationFriday': True,
        'programActivationMonday': True,
        'programActivationSaturday': True,
        'programActivationStatus': True,
        'programActivationSunday': True,
        'programActivationThursday': True,
        'programActivationTuesday': True,
        'programActivationWednesday': True,
        'programDepartureTime': '02:00',
        'programType': 'CHARGE_AND_PRECONDITIONING',
      }),
    ]),
  })
# ---
